rustbuild should compile rsbegin/rsend for TARGET not for HOST

Because that's what the old build system does:

```
LD_LIBRARY_PATH=/home/japaric/x86_64-unknown-linux-gnu/stage2/lib:/usr/lib:$LD_LIBRARY_PATH   x86_64-unknown-linux-gnu/stage2/bin/rustc --cfg stage2  -O --cfg rtopt -C rpath  --target=x86_64-pc-windows-gnu -C linker=x86_64-w64-mingw32-gcc -C ar=x86_64-w64-mingw32-ar   --emit=obj -o x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-pc-windows-gnu/lib/rsbegin.o /shared/rust/master/src/rtstartup/rsbegin.rs
```

but rustbuild [doesn't pass the `--target`](https://github.com/rust-lang/rust/blob/cbe4de78e2daf84499988c79317c0862b8106115/src/bootstrap/compile.rs#L124-L127) and rsbegin.o ends up being an ELF on linux:

```
$ file $(find -name rsbegin.o)
./build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-pc-windows-gnu/lib/rsbegin.o: ELF 64-bit LSB relocatable, x86-64, version 1 (SYSV), not stripped
```

The fix should be simple in theory. You just have to pass `--target $TARGET` to the `Command` in the code linked above. But then you run in problems with the order of the operations: "`core` not found for $TARGET".


### Meta

```
$ git rev-parse HEAD
e77d86c142ae668038dd43594d04022cbd6bf4d8
```

cc @alexcrichton

FWIW, trying to cross compile to **i686**-pc-windows-gnu I get this:

```
   Compiling std v0.0.0 (file:///shared/rust/master/src/libstd)
error: linking with `i686-w64-mingw32-gcc` failed: exit code: 1
  |
  = note: "i686-w64-mingw32-gcc" "-Wl,--enable-long-section-names" "-fno-use-linker-plugin" "-Wl,--nxcompat" "-nostdlib" "-Wl,--large-address-aware" "/home/japaric/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/i686-pc-windows-gnu/lib/dllcrt2.o" "/home/japaric/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/i686-pc-windows-gnu/lib/rsbegin.o" "-L" "/home/japaric/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/i686-pc-windows-gnu/lib" "/home/japaric/build/x86_64-unknown-linux-gnu/stage1-std/i686-pc-windows-gnu/release/deps/std-e2265790837cbaa9.0.o" "-o" "/home/japaric/build/x86_64-unknown-linux-gnu/stage1-std/i686-pc-windows-gnu/release/deps/std-e2265790837cbaa9.dll" "/home/japaric/build/x86_64-unknown-linux-gnu/stage1-std/i686-pc-windows-gnu/release/deps/std-e2265790837cbaa9.metadata.o" "-nodefaultlibs" "-L" "/home/japaric/build/x86_64-unknown-linux-gnu/stage1-std/i686-pc-windows-gnu/release/deps" "-L" "/home/japaric/build/x86_64-unknown-linux-gnu/stage1-std/i686-pc-windows-gnu/release/build/std-a76d8b43f160511e/out/.libs" "-L" "/home/japaric/build/x86_64-unknown-linux-gnu/stage1-std/i686-pc-windows-gnu/release/build/alloc_jemalloc-b4a9fdd2a74b7a98/out/lib" "-L" "/home/japaric/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/i686-pc-windows-gnu/lib" "-Wl,-Bstatic" "-Wl,--whole-archive" "-l" "backtrace" "-Wl,--no-whole-archive" "-Wl,-Bdynamic" "-l" "advapi32" "-l" "ws2_32" "-l" "userenv" "-l" "shell32" "-Wl,--whole-archive" "/tmp/rustc.EO8OdN3FN33s/libcollections-f1089b41bd63d58e.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc.EO8OdN3FN33s/libpanic_unwind-3976021b89fe79d0.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc.EO8OdN3FN33s/liballoc-bc4e10ff5ace48d6.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc.EO8OdN3FN33s/liballoc_system-0126d25e93e8a210.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc.EO8OdN3FN33s/librustc_unicode-42e6477b8bd9d820.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc.EO8OdN3FN33s/libunwind-7f3a35b0cd66ef0a.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc.EO8OdN3FN33s/liblibc-6f0a493cfbc94393.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc.EO8OdN3FN33s/librand-8de767ad0539476d.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc.EO8OdN3FN33s/libcore-bf8588192495524e.rlib" "-Wl,--no-whole-archive" "-l" "gcc_eh" "-l" "gcc_eh" "-shared" "-l" "compiler-rt" "-lmingwex" "-lmingw32" "-lgcc" "-lmsvcrt" "-luser32" "-lkernel32" "/home/japaric/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/i686-pc-windows-gnu/lib/rsend.o"
  = note: /home/japaric/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/i686-pc-windows-gnu/lib/rsbegin.o: file not recognized: File format not recognized
collect2: error: ld returned 1 exit status


error: aborting due to previous error

error: Could not compile `std`.
```

Trying to cross compile to **x86_64**-pc-windows-gnu results in #36291

--- a/src/bootstrap/compile.rs
+++ b/src/bootstrap/compile.rs
@@ -117,14 +117,16 @@ fn build_startup_objects(build: &Build, target: &str, into: &Path) {
         return
     }
     let compiler = Compiler::new(0, &build.config.build);
-    let compiler = build.compiler_path(&compiler);
+    let compiler_path = build.compiler_path(&compiler);
 
     for file in t!(fs::read_dir(build.src.join("src/rtstartup"))) {
         let file = t!(file);
-        build.run(Command::new(&compiler)
-                          .arg("--emit=obj")
-                          .arg("--out-dir").arg(into)
-                          .arg(file.path()));
+        let mut cmd = Command::new(&compiler_path);
+        build.add_bootstrap_key(&compiler, &mut cmd);
+        build.run(cmd.arg("--target").arg(target)
+                     .arg("--emit=obj")
+                     .arg("--out-dir").arg(into)
+                     .arg(file.path()));
     }
 
     for obj in ["crt2.o", "dllcrt2.o"].iter() {

--- a/src/rtstartup/rsbegin.rs
+++ b/src/rtstartup/rsbegin.rs
@@ -22,9 +22,18 @@
 // object (usually called `crtX.o), which then invokes initialization callbacks
 // of other runtime components (registered via yet another special image section).
 
+#![feature(no_core, lang_items)]
 #![crate_type="rlib"]
-#![no_std]
+#![no_core]
 #![allow(non_camel_case_types)]
+
+#[lang = "sized"]
+trait Sized {}
+#[lang = "sync"]
+trait Sync {}
+#[lang = "copy"]
+trait Copy {}
+impl<T> Sync for T {}
 
 #[cfg(all(target_os="windows", target_arch = "x86", target_env="gnu"))]
 pub mod eh_frames {

--- a/src/rtstartup/rsend.rs
+++ b/src/rtstartup/rsend.rs
@@ -10,8 +10,15 @@
 
 // See rsbegin.rs for details.
 
+#![feature(no_core, lang_items)]
 #![crate_type="rlib"]
-#![no_std]
+#![no_core]
+
+#[lang = "sized"]
+trait Sized {}
+#[lang = "sync"]
+trait Sync {}
+impl<T> Sync for T {}
 
 #[cfg(all(target_os="windows", target_arch = "x86", target_env="gnu"))]
 pub mod eh_frames {

--- a/src/rustc/std_shim/Cargo.toml
+++ b/src/rustc/std_shim/Cargo.toml
@@ -41,6 +41,7 @@ debug-assertions = false
 
 [dependencies]
 std = { path = "../../libstd" }
+core = { path = "../../libcore" }
 
 # Reexport features from std
 [features]

--- a/src/rustc/std_shim/lib.rs
+++ b/src/rustc/std_shim/lib.rs
@@ -9,3 +9,9 @@
 // except according to those terms.
 
 // See comments in Cargo.toml for why this exists
+
+// There's a bug right now where if we pass --extern std=... and we're cross
+// compiling then this doesn't work with `#[macro_use] extern crate std;`. Work
+// around this by not having `#[macro_use] extern crate std;`
+#![no_std]
+extern crate std;

